Format expression Android bindings

LukasPaczos opened the pull request:
Refs https://github.com/mapbox/mapbox-gl-native/pull/12624, exposes `format` expression on Android.

Setting the `text-field` using the expression, or plain text, works as expected, however, getting the `text-field` property back from core reveals a couple of issues:
 - Plain text is returned as expected.
 - Any expression, other than `format`, is returned wrapped by `format` expression. This includes tokens (which are beforehand converted from plain text to expressions and this is expected).
 - When value is passed initially as the `format` expression, it's being returned as `String`, which is expected because of [this implementation](https://github.com/mapbox/mapbox-gl-native/pull/12624/files#diff-dc7d81f6247193e64e085c48fc54e29cR33).

Currently, above issues cause a bunch of tests to fail.

/cc @ChrisLoer

LukasPaczos commented:
Solution to above issues will require introducing a `Formatted` type on the Java side, and possibly separate getters and setters of `text-field` property for plain text and `Expression` + `Formatted`.

LukasPaczos commented:
This PR introduces `Formatted` java object which is returned in, also newly added, `SymbolLayer#getFormattedTextField` value field for constant `text-field` properties. In order not to break semver, the `SymbolLayer#getTextField` value field returns plain text in the form of `String`.

Expression fields returned by `SymbolLayer#getFormattedTextField` and `SymbolLayer#getTextField` both return the original `Expression` wrapped in `Expression#format` expression. 

Follow up work needed after this PR lands is exposing `PropertyFactory#textField` method that takes `Formatted` as an argument and all the c++ conversion code needed. When major version release comes, we can remove `SymbolLayer#getFormattedTextField` and make `SymbolLayer#getTextField` return `Formatted` value type.

ChrisLoer commented on platform/android/src/style/formatted.cpp:
```diff
@@ -0,0 +1,42 @@
+            fontScale = section.fontScale.value();
+        }
+
+        auto fontStack = jni::Array<jni::String>::New(env, 0);
```
When I saw `FormattedSection::getFontStack()` above and saw that it was nullable, I assumed that "null" was being used to represent "not set". But it looks here (and now I notice in the tests) that `text-font: null` will cause `getFontStack()` to return `new String[] {}`. That doesn't seem right -- an empty array would signify "use this FontStack that doesn't have any fonts in it" instead of "use the default font stack".

LukasPaczos replied:
Good point, I'm going to return `null` instead of an empty array or un-set stack.

ChrisLoer commented on platform/android/src/style/formatted.cpp:
```diff
@@ -0,0 +1,42 @@
+    jni::Class<Formatted>::Singleton(env);
+}
+
+jni::Local<jni::Object<Formatted>> Formatted::New(jni::JNIEnv& env, const style::expression::Formatted& value) {
```
This is just an idea... but this is a lot of JNI code to kind of directly wrap the constituents of `expression::Formatted` in Java equivalents, and JNI bridging code is hard to audit and reason about (including the perf implications of going back and forth across the bridge multiple times). Would it maybe make sense to try an alternative approach where the bridging code just uses `stringify(Writer& writer, const expression::Formatted& v)` to turn the `expression::Formatted` into a JSON string, uses the standard string bridging code, and then parses the JSON on the java side? It's not at all clear to me this would be a win in terms of clarity or performance, but thought it might be worth considering...

LukasPaczos replied:
JSON parsing has been a bottleneck on multiple occasions and I feel like JNI overhead might be the lesser of two evils in this case.

ChrisLoer replied:
OK, fair enough. Since we're now doing this even for the string-typed `getTextField` (since the format->string flattening happens in java-land now), this could affect performance for existing code. Do you have an idea what an appropriate benchmark for this would be? I don't have a good feel for how often users will typically be hitting `getTextField`, it totally makes sense that it might not be nearly heavy enough to matter.

LukasPaczos replied:
I did a quick test - 1000 subsequent calls to `getTextField`:
- `master` ~6ms
- this PR ~21ms

The overhead is significant overall, but I don't think it's significant in this particular scenario. Constant `text-field` is definitely not a property that's going to be queried a lot as it depends on source updates that need to be performed by the user, as do all constant properties. Also, I still believe that JSON parsing on the java side would introduce even bigger performance hit.

ChrisLoer replied:
Sounds good. A ~4x slowdown doesn't sound disastrous in this context at all.

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/types/FormattedSection.java:
```diff
@@ -0,0 +1,84 @@
+   * @return font stack
+   */
+  @Nullable
+  public String[] getFontStack() {
```
atm we do, but definitely an improvement worth investing into

tobrun commented on platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/style/expressions/ExpressionTest.java:
```diff
@@ -1403,4 +1404,66 @@ public void testIsSupportedScriptExpressions() {
+      }
+    };
+    Object[] actual = format(
+      literal("test"), null, literal("awesome"),
```
`"awesome"` should be `new String[]{"awesome"}`
same applies for other occurrences of font-stack in the unit tests

LukasPaczos replied:
great catch!

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/layer.java.ejs:
```diff
@@ -18,6 +18,8 @@ import static com.mapbox.mapboxsdk.utils.ColorUtils.rgbaToColor;
 import com.google.gson.JsonArray;
 import com.mapbox.mapboxsdk.style.expressions.Expression;
 import com.mapbox.mapboxsdk.style.layers.TransitionOptions;
+import com.mapbox.mapboxsdk.style.types.Formatted;
```
Added the check.

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/expressions/Expression.java:
```diff
@@ -3224,6 +3226,78 @@ public static Expression collator(Expression caseSensitive, Expression diacritic
+   * "format" expressions can be used, for example, with the {@link PropertyFactory#textField(Expression)}
+   * and accepts unlimited numbers of formatted sections.
+   * <p>
+   * Each sections consist of 3 elements, in this order: text, font-scale and text-font.
```
Thanks for the heads up! I'll try to future-proof the builder.

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/types/FormattedSection.java:
```diff
@@ -0,0 +1,84 @@
+   * Creates a formatted section.
+   *
+   * @param text      displayed string
+   * @param fontScale scale of the font, 1.0 is default
```
This works as an equivalent but as you pointed out it might not be visible right away. Since this PR doesn't add an option to use `Formatted` as a `text-font` argument, I'm going to make the constructor package-private (as it's called only from JNI) for the time being and expose it in a follow-up PR, making it clearer at the same time.

LukasPaczos commented:
This one is ready for another round @tobrun @ChrisLoer. I addressed the comments and opted for a more typed builder with `FormatEntry` type suggested by @tobrun and extended the idea to include `FormatOption` which also prepares the code to accept a new set of options in the future.

LukasPaczos commented:
https://github.com/mapbox/mapbox-gl-native/pull/12985/commits/923dfa7f27ef4efb7f649b55faaece6cef952860 also updates the local style that we are using for testing which was causing issues when setting a `text-field` property.

LukasPaczos commented:
https://github.com/mapbox/mapbox-gl-native/pull/12985/commits/8395a5c835437c0cf414768e56d81e967d40dcfb renames format options to avoid a clash with `PropertyFactory#textFont` and adds the example in the test activity.

ChrisLoer commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/expressions/Expression.java:
```diff
@@ -3224,6 +3226,123 @@ public static Expression collator(Expression caseSensitive, Expression diacritic
+   *         formatFontScale(2.0),
+   *         formatTextFont(new String[] {"DIN Offc Pro Regular", "Arial Unicode MS Regular"})
+   *       ),
+   *       formatEntry(concat(literal("\n"), get("description_property")), formatFontScale(1.5)),
```
This code example looks like it abruptly ends after `formatFontScale(1.5)),`, is that intentional?

LukasPaczos replied:
Nice catch :+1:

ChrisLoer commented on platform/android/src/style/position.cpp:
```diff
@@ -31,5 +31,5 @@ float Position::getPolarAngle(jni::JNIEnv& env, const jni::Object<Position>& pos
     return position.Get(env, field);
 }
 
-} // namespace andr[oid
```
😆

ChrisLoer commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/testapp/style/SymbolLayerTest.java:
```diff
@@ -1004,7 +1005,6 @@ public void testIconHaloColorAsExpression() {
     });
   }
 
-
```
Nit: can we get rid of this diff? I looked at layer.junit.ejs and couldn't actually see where it was coming from..

LukasPaczos replied:
It's hard to tell which change is it, and since it actually fixes the formatting of the resulting files, I'm going to keep it in.

LukasPaczos commented:
Example usage of the `format` expression:
![device-2019-03-05-121105](https://user-images.githubusercontent.com/16925074/53801688-1a1bd800-3f40-11e9-8d7b-8f5086b8ef11.png)

```
mapView.getMapAsync(mapboxMap -> {
  mapboxMap.setStyle(Style.MAPBOX_STREETS, style -> {
    // create header entry based on data-driven styling
    Expression.FormatEntry header = Expression.formatEntry(
      Expression.get("header_property"),
      Expression.FormatOption.formatFontScale(Expression.get("header_scale_property")),
      Expression.FormatOption.formatTextFont(Expression.get("header_font_property")));
    
    // create hardcoded text entry
    Expression.FormatEntry text = Expression.formatEntry(
      "\nsome text",
      Expression.FormatOption.formatFontScale(1.5));
    
    // create hardcoded side-note entry
    Expression.FormatEntry sideNote = Expression.formatEntry(
      " (side-note)",
      Expression.FormatOption.formatFontScale(0.75));
    
    // build format expression out of provided entries
    Expression format = Expression.format(header, text, sideNote);
    
    // create the first feature and add header data
    Feature feature1 = Feature.fromGeometry(Point.fromLngLat(0, 45));
    feature1.addStringProperty("header_property", "Header1");
    feature1.addNumberProperty("header_scale_property", 3.5);
    
    JsonArray headerFontStack1 = new JsonArray();
    headerFontStack1.add("DIN Offc Pro Bold");
    headerFontStack1.add("Arial Unicode MS Regular");
    feature1.addProperty("header_font_property", headerFontStack1);
    
    // create the second feature and add header data
    Feature feature2 = Feature.fromGeometry(Point.fromLngLat(0, -45));
    feature2.addStringProperty("header_property", "Header2");
    feature2.addNumberProperty("header_scale_property", 3.5);
    
    // reverse the font stack
    JsonArray headerFontStack2 = new JsonArray();
    headerFontStack2.add("Arial Unicode MS Regular");
    headerFontStack2.add("DIN Offc Pro Bold");
    feature2.addProperty("header_font_property", headerFontStack2);
    
    // add the source and the layer
    Source source = new GeoJsonSource("source", FeatureCollection.fromFeatures(new Feature[] {feature1, feature2}));
    style.addSource(source);
    Layer layer = new SymbolLayer("layer", "source")
      .withProperties(
        PropertyFactory.textField(format)
      );
    style.addLayer(layer);
  });
});
```
